### PR TITLE
Support blob streaming for file-like objects

### PR DESCRIPTION
**Pull request by linar-jether**
When uploading file-like objects without a specified size, upload in chunks instead of exhausting the stream and keeping the data in-memory.

See https://github.com/GoogleCloudPlatform/gcloud-python/issues/1909

### CODE REVIEW

**Comment by linar-jether** [comment]
Signed it

File: gcloud/streaming/buffered_stream.py [hunk 1]
```diff
@@ -20,7 +20,9 @@ def __init__(self, stream, start, size):
         self._stream = stream
         self._start_pos = start
         self._buffer_pos = 0
-        self._buffered_data = self._stream.read(size)
+        self._buffered_data = self._stream.read(size) \
+            if not hasattr(self._stream,
+                           'closed') or not self._stream.closed else ''
```
Comment (tseaver) [inline gcloud/streaming/buffered_stream.py]: No backslash continuations, please.  This assignment would be easier to read as a normal `if`..`else` statement.

File: gcloud/storage/blob.py [hunk 2]
```diff
@@ -510,6 +511,13 @@ def upload_from_file(self, file_obj, rewind=False, size=None,
         if self.chunk_size is not None:
             upload.chunksize = self.chunk_size
 
+            if total_bytes is None:
+                upload.strategy = RESUMABLE_UPLOAD
+        elif self.chunk_size is None and total_bytes is None:
```
Comment (tseaver) [inline gcloud/storage/blob.py]: We know unconditionally that that `self.chunk_size` is None here (or we would have taken the prior branch).

**Comment by linar-jether** [comment]
fixed according to comments

**Comment by tseaver** [comment]
@linar-jether thanks for the changes.  The pylint failures on Travis aren't due to this PR:  they released a new version last Friday, which introduced new checks that we now fail.  The `master` branch now has fixes for those failures:  could you rebase your branch against it and re-push?

**Comment by linar-jether** [comment]
All tests passed

File: gcloud/storage/blob.py [hunk 3]
```diff
@@ -17,7 +17,7 @@
 import base64
 import copy
 import hashlib
-from io import BytesIO
+from io import BytesIO, UnsupportedOperation
```
Comment (daspecster) [inline gcloud/storage/blob.py]: We really try to keep it to one import per line.

```python
from io import BytesIO
from io import UnsupportedOperation
```

**Comment by daspecster** [comment]
@linar-jether, cool! Do you think you could squash these commits down?
Also, I noticed the imports issue.

Otherwise LGTM, but I defer to @tseaver/@dhermes for confirmation.

**Comment by linar-jether** [comment]
Sure, squashed all previous commits and separated the imports

**Comment by tseaver** [comment]
@linar-jether Thanks again for your work!

**Pull request merged**